File: LC/[1035] Uncrossed Lines.py
class Solution:
    def maxUncrossedLines(self, nums1: List[int], nums2: List[int]) -> int:
        """
        this is asking LCS
        N=500
        """
        nums1 = [0]+nums1
        nums2 = [0]+nums2
        N1 = len(nums1)
        N2 = len(nums2)
        dp = [[0]*N2 for _ in range(N1)]

        for ii in range(1, N1):
            for jj in range(1, N2):
                if nums1[ii] == nums2[jj]:
                    dp[ii][jj] = 1+dp[ii-1][jj-1]
                else:
                    dp[ii][jj] = max(dp[ii-1][jj], dp[ii][jj-1])

        return dp[-1][-1]

File: LC/[1035] Uncrossed Lines.cpp
class Solution {
    public:
        int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
            //this is same as LCS: longest common subsequence
            int N1 = nums1.size();
            int N2 = nums2.size();
    
            nums1.insert(nums1.begin(), 0);
            nums2.insert(nums2.begin(), 0);
    
            auto dp = vector<vector<int>>(N1+1, vector<int>(N2+1, 0));
            for(int ii=1; ii<=N1; ++ii){
                for(int jj=1; jj<=N2; ++jj){
                    if(nums1[ii]==nums2[jj]){
                        dp[ii][jj] = dp[ii-1][jj-1]+1;
                    }else{
                        dp[ii][jj] = max(dp[ii-1][jj], dp[ii][jj-1]);
                    }
                }
            }
    
    
            return dp[N1][N2];
        }
    };
    
    /*
    * two series
    * nums1:
    * nums2:
    */